### PR TITLE
Add shared message functionality

### PR DESCRIPTION
**Pull request by toolness**
Currently this uses [joined table inheritance](http://docs.sqlalchemy.org/en/rel_1_0/orm/inheritance.html#joined-table-inheritance) to define an `Event` model, which is a generic event that can appear in a deployment's event stream. A `SharedMessageEvent` is then subclassed from this, representing a message a user has shared with the rest of the network.

An alternate option is to use [single table inheritance](http://docs.sqlalchemy.org/en/rel_1_0/orm/inheritance.html#single-table-inheritance); I didn't use this because joined table inheritance *seems* more extensible and maintainable, but we can switch if needed.

### CODE REVIEW

File: app/models.py [hunk 1]
```diff
@@ -340,3 +340,50 @@ class UserSkill(db.Model): #pylint: disable=no-init,too-few-public-methods
...
+    '''
+
+    __tablename__ = 'events'
+    id = Column(types.Integer, autoincrement=True, primary_key=True)  #pylint: disable=invalid-name
+    created_at = Column(types.DateTime(), default=datetime.datetime.now)
+    updated_at = Column(types.DateTime(), default=datetime.datetime.now,
+                        onupdate=datetime.datetime.now)
+    type = Column(types.String)
+
+    user_id = Column(types.Integer, ForeignKey('users.id'))
```
Comment (talos) [inline app/models.py]: I'm a little wary of this.  I want to check in with Batu that we don't imagine anything ever happening on the stream that's not related to a user.

**Comment by talos** [comment]
I think you're on the right track.  We'll probably want to handle conditionals (different rendering based off of different even types) at the template level.

File: app/models.py (hunk 1 above)
Comment (toolness) [inline app/models.py]: Yeah, I'm not sure about it either... Another alternative is to have a subclass of `Event` called `UserEvent` which most (if not all) events subclass from. Though if we don't require that users be affiliated with events, we might have to add a `deployment` field to the `Event` model too.

Comment (toolness) [inline app/models.py]: er, also, I didn't specify `nullable` on the `user_id` field since I assumed `user_id` could be null, but that proves problematic too since there would be no way to figure out what deployment an event is for. So perhaps `deployment` *is* needed as an explicit field...

Comment (talos) [inline app/models.py]: Yeah, it's not the worst thing if it is.  We could do the `UserEvent` subclass you're talking about, and have `deployment` live on the top-level class.  We'd pre-populate it with the user's `deployment` on creation.

Comment (toolness) [inline app/models.py]: Ok, just changed things to work this way.

File: app/models.py [hunk 2]
```diff
@@ -339,3 +340,64 @@ class UserSkill(db.Model): #pylint: disable=no-init,too-few-public-methods
...
+        'polymorphic_identity': 'event',
+        'polymorphic_on': type
+    }
+
+    @classmethod
+    def query_in_deployment(cls):
+        '''
+        Query for events within this deployment
+        '''
+        return cls.query.filter(cls.deployment.in_(current_app.config['SEARCH_DEPLOYMENTS']))
```
Comment (toolness) [inline app/models.py]: Since the `deployment` field and `query_in_deployment` are both identical to that of the `User` model, it might be useful to create a `DeploymentMixin` that defines them, to keep our code DRY.

File: app/forms.py [hunk 3]
```diff
@@ -90,6 +90,16 @@ class Meta:  #pylint: disable=no-init,missing-docstring,old-style-class,too-few-
         choices=lambda: [(v, lazy_gettext(v)) for v in current_app.config['DOMAINS']])
 
 
+class SharedMessageForm(Form):
+    '''
+    Form for submittin a message to share with the network.
+    '''
+
+    message = TextAreaField(
+        label=lazy_gettext('Have a question or something to share about innovation?')
```
Comment (toolness) [inline app/forms.py]: Not entirely sure if a `<label>` will be the right place to put this text from the wireframes, but it should make it accessible to screen readers at least...

File: app/models.py [hunk 4]
```diff
@@ -103,12 +104,16 @@ def query_in_deployment(cls):
...
+        return u"%s %s" % (self.first_name, self.last_name)
+
+    @property
     def display_in_search(self):
         '''
         Determine whether user has filled out bare minimum to display in search
         results.
         '''
-        return self.first_name is not None and self.last_name is not None
+        return bool(self.first_name and self.last_name)
```
Comment (toolness) [inline app/models.py]: This method is funky because anyone who has submitted the profile page form and left the first/last name blank now has a *blank* first and last name, which is not the same as a NULL first/last name... I've rewritten this method to hopefully be more accurate but let me know if my assumptions are totally off.

File: app/views.py [hunk 5]
```diff
@@ -244,6 +246,34 @@ def recent_users():
...
+    '''
+    View for the activity feed of recent events.
+    '''
+
+    events = Event.query_in_deployment().order_by(desc(Event.created_at)).limit(10).all()
+    shared_message_form = SharedMessageForm()
+
+    if request.method == 'POST':
+        if not current_user.display_in_search:
+            flash(gettext(u'We need your name before you can post a message.'), 'error')
```
Comment (toolness) [inline app/views.py]: An alternative here is to add the message to the database but simply not display it if the user's name is blank... Or we could substitute "Mysterious Person" as their name in the activity feed! I suspect the number of mysterious people will plummet once we make first/last name part of the signup process anyways?

**Comment by toolness** [comment]
Does our model assume that everyone has a first and last name? What if Madonna or Sinbad signed up?

File: app/forms.py (hunk 3 above)
Comment (talos) [inline app/forms.py]: We can figure this out down the line.

File: app/models.py (hunk 4 above)
Comment (talos) [inline app/models.py]: This is more accurate -- not a huge fan of the syntax, so maybe update the docstring to say what this does (makes sure that first name last name are both non-null and non-blank)?

File: app/views.py (hunk 5 above)
Comment (talos) [inline app/views.py]: I think we should require first name/last name in signup -- if they don't provide it, they can't advance.

**Comment by talos** [comment]
No Madonna in NoI -- only Madonna Ciccone. ;)

File: app/models.py [hunk 6]
```diff
@@ -36,7 +37,7 @@ class User(db.Model, UserMixin): #pylint: disable=no-init,too-few-public-methods
 
     has_picture = Column(types.Boolean, default=False)
     deployment = Column(types.String, nullable=False,
-                        default=lambda: current_app.config['NOI_DEPLOY'])
+                        default=get_current_deployment)
```
Comment (talos) [inline app/models.py]: :+1:

File: app/models.py (hunk 2 above)
Comment (talos) [inline app/models.py]: I would agree.

File: app/views.py (hunk 5 above)
Comment (talos) [inline app/views.py]: More generally, I think this if/else statement and associated test aren't necessary --we should handle this error situation elsewhere, not at message creation time.

File: app/tests/test_views.py [hunk 7]
```diff
@@ -56,6 +56,34 @@ def login(self, email=None, password=None):
         assert LOGGED_IN_SENTINEL in res.data
         return res
 
+class ActivityFeedTests(ViewTestCase):
+    def test_posting_activity_requires_full_name(self):
```
Comment (talos) [inline app/tests/test_views.py]: I don't think we need to test for this.  We shouldn't be letting users in without names, and if somehow one does slip in, I think we should just hide their post.

File: app/views.py (hunk 5 above)
Comment (talos) [inline app/views.py]: Changed my mind.  Let's leave it.  Coding different edge-case behavior isn't worth the effort.

File: app/tests/test_views.py (hunk 7 above)
Comment (talos) [inline app/tests/test_views.py]: Let's just leave this in, it's a better solution than spending more time dealing with this edge case.

**Comment by talos** [comment]
Great work! :boom:
